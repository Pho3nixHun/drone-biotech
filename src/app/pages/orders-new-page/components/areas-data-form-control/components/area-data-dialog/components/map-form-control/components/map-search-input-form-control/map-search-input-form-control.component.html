<div
    class="w-full relative"
    *transloco="let t"
    appClickOutside
    [attachedCallback]="close"
>
    <div class="flex flex-row w-full">
        <input
            type="text"
            appElementRef
            [signal]="searchInputSignal"
            [formControl]="control"
            [placeholder]="t(vm().placeholderKey)"
            (focus)="open()"
            (keyup.arrowup)="onArrowUp()"
            (keyup.arrowdown)="onArrowDown()"
            (keydown.enter)="
                selectPlace(this.placeSearchResults()[this.index()])
            "
            class="w-full bg-white p-4 h-12 border-r-0 border"
        />
        <div class="h-12 bg-white flex items-center px-4 border-l-0 border">
            <mat-icon>search</mat-icon>
        </div>
    </div>

    @if (isOpened() && placeSearchResults().length > 0) {
        <div class="absolute w-full z-10 bg-white border border-t-0">
            @for (place of placeSearchResults(); track $index) {
                <button
                    class="flex flex-row focus:outline-none border-none hover:bg-gray-200 w-full h-full px-5 py-2 justify-between"
                    [ngClass]="$index === index() ? 'bg-gray-200' : 'bg-white'"
                    (click)="selectPlace(place)"
                >
                    <h2 class="text-base font-normal">
                        {{ place.description }}
                    </h2>
                    @if (place.distance) {
                        <p class="font-thin">
                            {{
                                t(vm().distanceValueKey, {
                                    value: place.distance,
                                })
                            }}
                        </p>
                    }
                </button>
            }
        </div>
    }
</div>
