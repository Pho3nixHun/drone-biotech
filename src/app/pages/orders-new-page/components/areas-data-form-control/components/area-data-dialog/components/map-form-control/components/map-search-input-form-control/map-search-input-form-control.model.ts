export interface MapSearchInputFormControlVM {
    placeholderKey: string;
    distanceValueKey: string;
}
