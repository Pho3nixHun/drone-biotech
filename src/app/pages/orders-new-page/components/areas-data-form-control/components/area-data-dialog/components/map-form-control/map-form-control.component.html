<div class="flex flex-col space-y-2">
    <app-map-search-input-form-control
        [defaultCenter]="vm().defaultCenter"
        [vm]="vm().mapSearchInputFormControlVM"
        (changeBounds)="setMapBounds($event)"
    />

    <div appElementRef [signal]="divSignal" style="height: 600px"></div>

    <app-map-area-select-form-control
        [mapRef]="mapSignal()"
        [vm]="vm().mapAreaSelectFormControlVM"
        [formControl]="mapForm.controls.targetArea"
    />
    <app-map-point-select-form-control
        [mapRef]="mapSignal()"
        [vm]="vm().mapPointSelectFormControlVM"
        [formControl]="mapForm.controls.entryPoint"
    />
</div>
