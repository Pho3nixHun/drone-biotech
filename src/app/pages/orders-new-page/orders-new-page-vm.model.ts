import { WithTitle } from '@interfaces/with-title.interface';
import { InputTextXVM } from '@components/input-text/input-text.component';
import { AreasDataFormControlVM } from './components/areas-data-form-control/areas-data-form-control.model';
import { ButtonXVM } from '@components/button/button.model';

interface AreasDataFormControlXVM extends AreasDataFormControlVM {
    addButtonXVM: ButtonXVM;
}

export interface OrdersNewPageVM extends WithTitle {
    areasDataFormControlXVM: AreasDataFormControlXVM;
    requiredAssistiveTextKey: string;
    internalOrderNumberInputTextXVM: InputTextXVM;
    nameInputTextXVM: InputTextXVM;
    phoneNumberInputTextXVM: InputTextXVM;
    emailInputTextXVM: InputTextXVM;
    endCustomerInputTextXVM: InputTextXVM;
    submitButtonXVM: ButtonXVM;
}
