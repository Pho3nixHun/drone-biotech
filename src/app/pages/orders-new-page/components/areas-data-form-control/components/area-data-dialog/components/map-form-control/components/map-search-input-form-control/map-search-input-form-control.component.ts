/* eslint-disable @typescript-eslint/no-empty-function */
import {
    Component,
    effect,
    ElementRef,
    forwardRef,
    inject,
    input,
    output,
    signal,
} from '@angular/core';
import { rxResource, toSignal } from '@angular/core/rxjs-interop';
import {
    ControlValueAccessor,
    FormBuilder,
    NG_VALUE_ACCESSOR,
    ReactiveFormsModule,
} from '@angular/forms';
import { MatIconModule } from '@angular/material/icon';
import { AutoCompleteService } from '@services/auto-complete/auto-complete.service';
import {
    debounceTime,
    distinctUntilChanged,
    filter,
    of,
    switchMap,
} from 'rxjs';
import { MapSearchInputFormControlVM } from './map-search-input-form-control.model';
import { Coordinates } from '@stores/location/location.model';
import { Place } from '@services/auto-complete/auto-complete.model';
import { TranslocoModule } from '@jsverse/transloco';
import { isNotNull } from '@utils/is-null.typeguard';
import { NgClass } from '@angular/common';
import { ElementRefDirective } from '@directives/element-ref/element-ref.directive';
import { ClickOutsideDirective } from '@directives/click-outside/click-outside.directive';

const noop = () => {};

@Component({
    selector: 'app-map-search-input-form-control',
    imports: [
        ReactiveFormsModule,
        MatIconModule,
        TranslocoModule,
        NgClass,
        ElementRefDirective,
        ClickOutsideDirective,
    ],
    providers: [
        {
            provide: NG_VALUE_ACCESSOR,
            useExisting: forwardRef(() => MapSearchInputFormControlComponent),
            multi: true,
        },
    ],
    templateUrl: './map-search-input-form-control.component.html',
})
export class MapSearchInputFormControlComponent
    implements ControlValueAccessor
{
    private readonly service = inject(AutoCompleteService);
    private readonly fb = inject(FormBuilder);

    public readonly vm = input.required<MapSearchInputFormControlVM>();
    public readonly defaultCenter = input.required<Coordinates | null>();
    public readonly changeBounds = output<google.maps.LatLngBounds | null>();

    protected readonly selectedPlace = signal<Place | null>(null);
    protected readonly isOpened = signal<boolean>(false);
    protected readonly index = signal<number>(0);
    protected readonly searchInputSignal =
        signal<ElementRef<HTMLInputElement> | null>(null);

    protected readonly control = this.fb.control('');

    protected readonly placeSearchResults = toSignal(
        this.control.valueChanges.pipe(
            debounceTime(300),
            distinctUntilChanged(),
            filter(isNotNull),
            switchMap((searchTerm) =>
                searchTerm.length > 0
                    ? this.service.getPlaces(searchTerm, this.defaultCenter())
                    : of([])
            )
        ),
        { initialValue: [] }
    );

    private readonly locationResource = rxResource({
        request: () => this.selectedPlace(),
        loader: ({ request }) =>
            request
                ? this.service.getLocationByPlaceId(request.placeId)
                : of(null),
        defaultValue: null,
    });

    private readonly selectedPlaceEffect = effect(() => {
        const place = this.locationResource.value();
        this.changeBounds.emit(place);
        this.onChange(place);
    });

    protected selectPlace(place: Place) {
        this.control.setValue(place.description);
        this.selectedPlace.set(place);
        this.searchInputSignal()?.nativeElement.blur();
        this.index.set(0);
        this.isOpened.set(false);
    }

    protected onArrowUp() {
        const index = this.index();
        if (index === 0) return;
        this.index.set(index - 1);
    }
    protected onArrowDown() {
        const index = this.index();
        const length = this.placeSearchResults().length;
        if (index + 1 === length) return;
        this.index.set(index + 1);
    }

    protected open() {
        this.isOpened.set(true);
        this.onTouched();
    }

    protected close = () => this.isOpened.set(false);

    public writeValue(place: Place) {
        this.selectedPlace.set(place);
    }

    private onChange: (value: google.maps.LatLngBounds | null) => void = noop;
    private onTouched: () => void = noop;

    public registerOnChange(
        fn: (value: google.maps.LatLngBounds | null) => void
    ): void {
        this.onChange = fn;
    }

    public registerOnTouched(fn: () => void): void {
        this.onTouched = fn;
    }
}
