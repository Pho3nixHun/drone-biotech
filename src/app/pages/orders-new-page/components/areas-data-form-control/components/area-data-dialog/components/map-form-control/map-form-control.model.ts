import { MapSearchInputFormControlVM } from 'src/app/pages/orders-new-page/components/areas-data-form-control/components/area-data-dialog/components/map-form-control/components/map-search-input-form-control/map-search-input-form-control.model';
import { MapAreaSelectFormControlVM } from './components/map-area-select-form-control/map-area-select-form-control.model';
import { MapPointSelectFormControlVM } from './components/map-point-select-form-control/map-point-select-form-control.model';
import { Coordinates } from '../../area-data-dialog.model';
import { InjectionToken } from '@angular/core';

export interface MapFormControlVM {
    defaultCenter: Coordinates | null;
    mapAreaSelectFormControlVM: MapAreaSelectFormControlVM;
    mapPointSelectFormControlVM: MapPointSelectFormControlVM;
    mapSearchInputFormControlVM: MapSearchInputFormControlVM;
}

export const MAP_OPTIONS = new InjectionToken<MapOptions>(
    'Injection Token for Map options'
);

export const provideMockMapOptions = () => ({
    provide: MAP_OPTIONS,
    useValue: {},
});
export interface MapOptions {
    mapTypeId: 'roadmap';
    disableDoubleClickZoom: boolean;
    clickableIcons: boolean;
    mapId: 'DEMO_MAP_ID';
    zoom: number;
    streetViewControl: boolean;
}

export const mapStringToCoordinates = (input: string): Coordinates[] => {
    const lines = input
        .split('\n')
        .map((line) => line.trim())
        .filter((line) => line !== ''); // ignore empty lines

    const coordinates: Coordinates[] = lines.map((line) => {
        const [lat, lng] = line.split(' ').map(Number);
        return { lat, lng };
    });

    return coordinates;
};

export const mapCoordinatesToString = (coordinates: Coordinates[]): string => {
    return coordinates.map((coord) => `${coord.lat} ${coord.lng}`).join('\n');
};
