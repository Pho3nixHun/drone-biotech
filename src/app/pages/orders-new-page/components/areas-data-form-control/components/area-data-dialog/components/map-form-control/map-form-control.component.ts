/* eslint-disable @typescript-eslint/no-empty-function */
import {
    Component,
    computed,
    effect,
    ElementRef,
    forwardRef,
    inject,
    input,
    Signal,
    signal,
} from '@angular/core';
import {
    ControlValueAccessor,
    FormBuilder,
    NG_VALUE_ACCESSOR,
    ReactiveFormsModule,
    Validators,
} from '@angular/forms';
import { MapFormControlVM } from './map-form-control.model';
import { Coordinates } from '@stores/location/location.model';
import { MapAreaSelectFormControlComponent } from './components/map-area-select-form-control/map-area-select-form-control.component';
import { MapPointSelectFormControlComponent } from './components/map-point-select-form-control/map-point-select-form-control.component';
import { MapSearchInputFormControlComponent } from './components/map-search-input-form-control/map-search-input-form-control.component';
import { distinctUntilChanged, map } from 'rxjs';
import { toSignal } from '@angular/core/rxjs-interop';
import { getCenter } from 'geolib';
import { ElementRefDirective } from '@directives/element-ref/element-ref.directive';
import {
    MAP_OPTIONS,
    HEAD_OFFICE_LOCATION,
} from '@providers/google-maps-provider';

@Component({
    selector: 'app-map-form-control',
    imports: [
        MapAreaSelectFormControlComponent,
        MapPointSelectFormControlComponent,
        ElementRefDirective,
        ReactiveFormsModule,
        MapSearchInputFormControlComponent,
    ],
    providers: [
        {
            provide: NG_VALUE_ACCESSOR,
            useExisting: forwardRef(() => MapFormControlComponent),
            multi: true,
        },
    ],
    schemas: [],
    templateUrl: './map-form-control.component.html',
})
export class MapFormControlComponent implements ControlValueAccessor {
    private readonly fb = inject(FormBuilder);
    private readonly mapOptions = inject(MAP_OPTIONS);
    protected readonly headOfficeLocation = inject(HEAD_OFFICE_LOCATION);
    public readonly vm = input.required<MapFormControlVM>();
    protected readonly divSignal = signal<ElementRef<HTMLElement> | null>(null);
    private readonly targetArea = signal<Coordinates[] | null>(null);

    protected readonly mapSignal = computed(() => {
        const canvas = this.divSignal();
        if (!canvas) return null;
        return new google.maps.Map(canvas.nativeElement);
    });

    private readonly targetAreaCenter = computed<Coordinates | null>(() => {
        const area = this.targetArea();
        if (!area) return null;
        const center = getCenter(area);
        return center
            ? {
                  lat: center.latitude,
                  lng: center.longitude,
              }
            : null;
    });

    protected readonly mapCenter = computed<Coordinates>(
        () =>
            this.targetAreaCenter() ??
            this.vm().defaultCenter ??
            this.headOfficeLocation
    );

    protected readonly setMapCenterEffect = effect(
        () => {
            const map = this.mapSignal();
            if (!map) return;
            map.setCenter(this.mapCenter());
            this.setMapCenterEffect.destroy();
        },
        { manualCleanup: true }
    );

    protected mapForm = this.fb.group({
        targetArea: this.fb.control<Coordinates[] | null>(
            null,
            Validators.required
        ),
        entryPoint: this.fb.control<Coordinates | null>(
            null,
            Validators.required
        ),
    });

    private readonly setMapOptionsEffect = effect(() => {
        const map = this.mapSignal();
        if (!map) return;

        map.setOptions({
            ...this.mapOptions,
            mapTypeControlOptions: {
                position: google.maps.ControlPosition.TOP_RIGHT,
            },
        });
    });

    protected setMapBounds(bounds: google.maps.LatLngBounds | null) {
        const map = this.mapSignal();
        if (!bounds || !map) return;
        map.fitBounds(bounds);
    }

    private readonly mapFormValue: Signal<{
        targetArea: Coordinates[] | null;
        entryPoint: Coordinates | null;
    }> = toSignal(
        this.mapForm.valueChanges.pipe(
            distinctUntilChanged(),
            map((data) => {
                const { entryPoint, targetArea } = data;
                return {
                    entryPoint: entryPoint ?? null,
                    targetArea: targetArea ?? null,
                };
            })
        ),
        { initialValue: { targetArea: null, entryPoint: null } }
    );

    private readonly mapFormValueEffect = effect(() =>
        this.onChange(this.mapFormValue())
    );

    private onChange: (value: {
        targetArea: Coordinates[] | null;
        entryPoint: Coordinates | null;
    }) => void = () => {};

    private onTouched: () => void = () => {};

    public writeValue(value: {
        targetArea: Coordinates[] | null;
        entryPoint: Coordinates | null;
    }) {
        this.mapForm.setValue(value);
        this.targetArea.set(value.targetArea);
    }

    public registerOnChange(fn: () => void): void {
        this.onChange = fn;
    }

    public registerOnTouched(fn: () => void): void {
        this.onTouched = fn;
    }
}
