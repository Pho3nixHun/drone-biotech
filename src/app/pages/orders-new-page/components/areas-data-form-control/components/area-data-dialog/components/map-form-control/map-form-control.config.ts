import { EnvironmentProviders, Provider } from '@angular/core';
import { HEAD_OFFICE_LOCATION } from '@services/distance/distance.model';
import {
    GOOGLE_MAPS_LOADER_CONFIG,
    provideGoogleMapsLibraries,
} from 'src/app/shared/providers/google-maps-provider';
import { environment } from 'src/environments/environment';
import {
    POLYGON_OPTIONS,
    INFO_WINDOW_OPTIONS,
} from './components/map-area-select-form-control/map-area-select-form-control.model';
import { ENTRY_POINT_MARKER_OPTIONS } from './components/map-point-select-form-control/map-point-select-form-control.model';
import { MAP_OPTIONS } from './map-form-control.model';

export const mapFormControlConfig: (Provider | EnvironmentProviders)[] = [
    provideGoogleMapsLibraries(),
    {
        provide: POLYGON_OPTIONS,
        useValue: {
            clickable: false,
            draggable: false,
            editable: true,
            fillOpacity: 0.5,
            strokeColor: 'blue',
            fillColor: 'blue',
        },
    },
    {
        provide: MAP_OPTIONS,
        useValue: {
            mapTypeId: 'roadmap',
            disableDoubleClickZoom: true,
            clickableIcons: false,
            mapId: 'DEMO_MAP_ID',
            zoom: 14,
            streetViewControl: false,
        },
    },
    {
        provide: ENTRY_POINT_MARKER_OPTIONS,
        useValue: {
            gmpDraggable: true,
        },
    },
    {
        provide: GOOGLE_MAPS_LOADER_CONFIG,
        useValue: {
            url: 'https://maps.googleapis.com/maps/api/js?',
            key: environment.googleMapsConfig.apiKey,
            version: '3.59',
            region: 'HU',
        },
    },
    {
        provide: HEAD_OFFICE_LOCATION,
        useValue: { lat: 47.312498121576795, lng: 21.309304570654604 },
    },
    {
        provide: INFO_WINDOW_OPTIONS,
        useValue: {
            disableAutoPan: true,
            headerDisabled: true,
        },
    },
];
