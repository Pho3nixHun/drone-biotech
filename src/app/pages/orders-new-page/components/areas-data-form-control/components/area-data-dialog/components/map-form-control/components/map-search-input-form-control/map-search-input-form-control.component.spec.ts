import {
    ComponentFixture,
    fakeAsync,
    TestBed,
    tick,
} from '@angular/core/testing';
import { provideMockMapOptions } from '../../map-form-control.model';
import { getTranslocoModule } from 'transloco-testing.module';
import { Component, DebugElement, inject, input, signal } from '@angular/core';
import { MapSearchInputFormControlVM } from './map-search-input-form-control.model';
import { MapSearchInputFormControlComponent } from './map-search-input-form-control.component';
import { enMock } from 'src/app/pages/orders-new-page/orders-new-page.mock';
import { FormBuilder, ReactiveFormsModule, Validators } from '@angular/forms';
import { Place } from '@services/auto-complete/auto-complete.model';
import { By } from '@angular/platform-browser';
import { provideAutoCompleteServiceMock } from '@services/auto-complete/auto-complete.service.mock';
import { Coordinates } from '@stores/location/location.model';

const en = {
    placeholder: 'place',
    distanceValue: 'distance',
};

const mockVM = {
    placeholderKey: en.placeholder,
    distanceValueKey: en.distanceValue,
};

const center = { lat: 10, lng: 10 };

const place = {
    description: 'description',
    distance: 1000,
    placeId: 'id',
};

// eslint-disable-next-line @typescript-eslint/no-empty-function
const noop = () => {};

@Component({
    imports: [MapSearchInputFormControlComponent, ReactiveFormsModule],
    template: `
        <app-map-search-input-form-control
            [vm]="vm()"
            [defaultCenter]="defaultCenter()"
            [formControl]="form"
            (changeBounds)="setMapBounds($event)"
        />
    `,
})
class TestHostComponent {
    private readonly fb = inject(FormBuilder);
    public vm = input.required<MapSearchInputFormControlVM>();
    public defaultCenter = input.required<Coordinates>();
    public bounds = signal<google.maps.LatLng | null>(null);
    public form = this.fb.control<Place | null>(null, Validators.required);
    public setMapBounds(bounds: google.maps.LatLng) {
        this.bounds.set(bounds);
    }
}

describe('MapSearchInputFormControlComponent', () => {
    let mapSearchInputFormControlComponent: MapSearchInputFormControlComponent;
    let component: TestHostComponent;
    let fixture: ComponentFixture<TestHostComponent>;
    let compiled: HTMLElement;
    let innerDebugElement: DebugElement;

    beforeEach(async () => {
        await TestBed.configureTestingModule({
            imports: [
                TestHostComponent,
                getTranslocoModule({
                    langs: { en: enMock },
                    translocoConfig: {
                        availableLangs: ['en'],
                        defaultLang: 'en',
                    },
                }),
            ],
            providers: [
                provideMockMapOptions(),
                provideAutoCompleteServiceMock(),
            ],
        }).compileComponents();

        fixture = TestBed.createComponent(TestHostComponent);
        fixture.componentRef.setInput('defaultCenter', center);
        fixture.componentRef.setInput('vm', mockVM);
        component = fixture.componentInstance;
        innerDebugElement = fixture.debugElement.query(
            By.directive(MapSearchInputFormControlComponent)
        );
        mapSearchInputFormControlComponent =
            innerDebugElement.componentInstance;
        compiled = innerDebugElement.nativeElement;
    });

    // Snapshot testing
    it('should render the template correctly when the VM is provided', () => {
        // Arrange

        // Act
        fixture.detectChanges();

        // Assert
        expect(compiled).toMatchSnapshot();
    });

    // Unit testing
    it('should write a null value to the CVA', () => {
        // There is no need to arrange

        // There is no need to act

        // Assert
        expect(mapSearchInputFormControlComponent['selectedPlace']()).toBe(
            null
        );
    });

    // Unit testing
    it('should write a Place value to the CVA', () => {
        // Arrange
        const val = {
            description: 'desc',
            distance: 100,
            placeId: 'id',
        };
        component.form.setValue(val);

        // Act
        fixture.detectChanges();

        // Assert
        expect(mapSearchInputFormControlComponent['selectedPlace']()).toBe(val);
    });

    // Snapshot testing
    xit(
        'should render the search options after the control value changes',
        noop
    );

    // Unit testing
    xit('should fetch the location of the place with the service and emit it via output if there is a selected place', async () => {
        // Arrange
        mapSearchInputFormControlComponent['selectedPlace'].set({
            description: 'Budapest',
            distance: 1000,
            placeId: 'id',
        });

        await fixture.whenStable();
    });

    it('should set the value of the search input signal by the ElementRefDirective', () => {
        // There is no need to arrange

        // Act
        fixture.detectChanges();

        // Assert
        expect(
            typeof mapSearchInputFormControlComponent['searchInputSignal']()
        ).toBe('object');
    });

    it('should the selectPlace(place) function set the control value, index, isOpened, selectedPlace', () => {
        // Arrange
        fixture.detectChanges();
        const searchInput =
            mapSearchInputFormControlComponent['searchInputSignal']()
                ?.nativeElement;
        const blurSpy = searchInput ? jest.spyOn(searchInput, 'blur') : null;
        mapSearchInputFormControlComponent['selectPlace'](place);

        // There is no need to act

        // Assert
        expect(mapSearchInputFormControlComponent['control'].value).toBe(
            place.description
        );
        expect(mapSearchInputFormControlComponent['index']()).toBe(0);
        expect(mapSearchInputFormControlComponent['isOpened']()).toBe(false);
        expect(mapSearchInputFormControlComponent['selectedPlace']()).toBe(
            place
        );
        expect(typeof blurSpy).toBe('function');
        if (blurSpy) expect(blurSpy).toHaveBeenCalled();
    });

    describe('when form is set with a valid value', () => {
        beforeEach(fakeAsync(() => {
            mapSearchInputFormControlComponent['control'].setValue('budapest');
            fixture.detectChanges();
            tick(300);
        }));

        // Unit testing
        it('should fetch the places with the service', fakeAsync(() => {
            // There is no need to arrange

            // There is no need to act

            // Assert
            expect(
                mapSearchInputFormControlComponent['placeSearchResults']()
            ).toStrictEqual([
                { description: 'Budapest', distance: 23, placeId: 'id1' },
                {
                    description: 'Budapest, Parlament',
                    distance: 25,
                    placeId: 'id2',
                },
            ]);
        }));

        describe('when the user interacts with the input', () => {
            let searchInput: HTMLInputElement | undefined;
            let arrowDown: KeyboardEvent;
            let arrowUp: KeyboardEvent;
            let enter: KeyboardEvent;
            beforeEach(() => {
                // Arrange
                searchInput =
                    mapSearchInputFormControlComponent['searchInputSignal']()
                        ?.nativeElement;
                arrowDown = new KeyboardEvent('keyup', {
                    key: 'ArrowDown',
                });
                arrowUp = new KeyboardEvent('keyup', {
                    key: 'ArrowUp',
                });
                enter = new KeyboardEvent('keydown', {
                    key: 'Enter',
                });
            });

            // Unit testing
            it('should set the place after push the Enter key on the input field', () => {
                if (searchInput) {
                    // Act
                    searchInput.focus();
                    searchInput.dispatchEvent(enter);

                    // Assert
                    expect(
                        mapSearchInputFormControlComponent['selectedPlace']()
                    ).toStrictEqual({
                        description: 'Budapest',
                        distance: 23,
                        placeId: 'id1',
                    });
                }
            });

            // Unit testing
            it('should set the index after navigate with the ArrowDown key', () => {
                if (searchInput) {
                    // Act
                    searchInput.focus();
                    searchInput.dispatchEvent(arrowDown);

                    // Assert
                    expect(mapSearchInputFormControlComponent['index']()).toBe(
                        1
                    );
                }
            });

            // Unit testing
            it('should set the index after navigate with the ArrowUp key', () => {
                if (searchInput) {
                    // Act
                    searchInput.focus();
                    searchInput.dispatchEvent(arrowDown);
                    searchInput.dispatchEvent(arrowUp);

                    // Assert
                    expect(mapSearchInputFormControlComponent['index']()).toBe(
                        0
                    );
                }
            });

            // Unit testing
            it('should set the index with 0 value at least after navigate with the ArrowUp key', () => {
                if (searchInput) {
                    // Act
                    searchInput.focus();
                    searchInput.dispatchEvent(arrowUp);
                    searchInput.dispatchEvent(arrowUp);

                    // Assert
                    expect(mapSearchInputFormControlComponent['index']()).toBe(
                        0
                    );
                }
            });
            // Unit testing
            it('should set the index with length of the searchPlaceResults at max after navigate with the ArrowUp key', () => {
                if (searchInput) {
                    // Act
                    searchInput.focus();
                    searchInput.dispatchEvent(arrowDown);
                    searchInput.dispatchEvent(arrowDown);
                    searchInput.dispatchEvent(arrowDown);

                    // Assert
                    expect(mapSearchInputFormControlComponent['index']()).toBe(
                        1
                    );
                }
            });
        });
    });
});
