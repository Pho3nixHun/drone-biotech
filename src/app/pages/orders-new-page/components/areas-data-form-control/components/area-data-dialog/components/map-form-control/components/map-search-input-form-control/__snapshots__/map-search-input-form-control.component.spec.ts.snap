// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapSearchInputFormControlComponent should render the template correctly when the VM is provided 1`] = `
<app-map-search-input-form-control
  class="ng-untouched ng-dirty ng-invalid"
  ng-reflect-default-center="[object Object]"
  ng-reflect-form="[object Object]"
  ng-reflect-vm="[object Object]"
>
  <div
    appclickoutside=""
    class="w-full relative"
    ng-reflect-attached-callback="() => this.isOpened.set(false)"
  >
    <div
      class="flex flex-row w-full"
    >
      <input
        appelementref=""
        class="w-full bg-white p-4 h-12 border-r-0 border ng-untouched ng-pristine ng-valid"
        ng-reflect-form="[object Object]"
        ng-reflect-signal="[Signal: null]"
        placeholder="place"
        type="text"
      />
      <div
        class="h-12 bg-white flex items-center px-4 border-l-0 border"
      >
        <mat-icon
          aria-hidden="true"
          class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
          data-mat-icon-type="font"
          role="img"
        >
          search
        </mat-icon>
      </div>
    </div>
    <!--container-->
  </div>
  <!--container-->
</app-map-search-input-form-control>
`;
